### PR TITLE
[TL] initial implement flashattention op in TL

### PR DESCRIPTION
**Pull request by tzj-fxz**
* add initial implementation of flashattention in ops/general_flashatten
* wrap implementation in config class and op class
* fix kernel name bug
* modify tilelang test script
* add ops flashattention test script

### CODE REVIEW

File: bitblas/ops/general_flashatten/__init__.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+def is_native_compute(Q_dtype, K_dtype, V_dtype) -> bool:
+    return Q_dtype == K_dtype and K_dtype == V_dtype
+
+
+@dataclass(frozen=True)
+class FlashAttenConfig(OperatorConfig):
+    batch: Union[int, Tuple[int]] = None
+    # TODO should distinguish from q_heads and kv_heads
+    heads: int = None
```
Comment (LeiWang1999) [inline bitblas/ops/general_flashatten/__init__.py:28]: recommend to be 

```
from typing import Optional
...
heads: Optional[int] = None
heads: Optional[int] = None
```

File: bitblas/ops/operator.py [hunk 2]
```diff
@@ -50,26 +51,36 @@ def __init__(self, config: OperatorConfig):
...
         self.config = config
+        self.kernel_name = ""
 
     @abstractmethod
     def is_valid_config(self, config: OperatorConfig):
         pass
 
     @abstractmethod
-    def generate(self, hint: Hint = None) -> str:
+    def _generate(self, hint: Hint = None) -> str:
```
Comment (LeiWang1999) [inline bitblas/ops/operator.py:61]: it’s worth considering if we should add a leading underline before `generate`?

In python function, `_` begin method might represent the function is `protected`.

File: bitblas/ops/operator.py [hunk 3]
```diff
@@ -50,26 +51,36 @@ def __init__(self, config: OperatorConfig):
         assert self.is_valid_config(config), (f"Invalid config for {self.__class__.__name__}: "
                                               f"{config}")
         self.config = config
+        self.kernel_name = ""
```
Comment (LeiWang1999) [inline bitblas/ops/operator.py:54]: The name of this value `kernel_name` is a bit ugly.

Should `BaseKernelNameGenerator` class keeps a `kernel_name` attribute?

**Review by LeiWang1999 — COMMENTED** [review]
Thanks for your contributions. @tzj-fxz 
I've left some comments.

File: bitblas/ops/operator.py (hunk 3 above)
Comment (tzj-fxz) [reply bitblas/ops/operator.py:54]: Thanks, this attribute has been replaced with a variable in "generate_valid" function

File: bitblas/ops/operator.py (hunk 2 above)
Comment (tzj-fxz) [reply bitblas/ops/operator.py:61]: A new function "generate_valid" will be the new interface for kernel name generate, with the call to "_generate" first and check the validity of the generated kernel name.
The protected "_generate" can be overwritten by sub class for real generation process.

File: bitblas/ops/general_flashatten/__init__.py (hunk 1 above)
Comment (tzj-fxz) [reply bitblas/ops/general_flashatten/__init__.py:28]: Thanks, both flashatten config and matmul config are modified using `Optional`

**Comment by LeiWang1999** [comment]
thanks @tzj-fxz ，overall lgtm, should fix the test.

**Comment by tzj-fxz** [comment]
> thanks @tzj-fxz ，overall lgtm, should fix the test.

OK, now CI tests are passed in the newest commit.

**Comment by LeiWang1999** [comment]
Merged, thanks @tzj-fxz

**Pull request merged**